src/java/org/apache/cassandra/db/ColumnFamilyStore.java

```diff
@@ -710,36 +721,64 @@ public static void removeUnfinishedCompactionLeftovers(CFMetaData metadata, Map<
                 assert compactionTaskID != null;
-                logger.info("Going to delete unfinished compaction product {}", desc);
-                SSTable.delete(desc, sstableFiles.getValue());
-                SystemKeyspace.finishCompaction(compactionTaskID);
+                logger.info("Going to delete unfinished compaction product", UnsafeArg.of("desc", desc),
+                            SafeArg.of("keyspace", desc.ksname), SafeArg.of("cf", desc.cfname),
+                            SafeArg.of("generation", desc.generation), ancestorsArg);
+                SSTable.delete(desc, allNonTempSstableFiles.get(desc));
+                cleanedUnfinishedCompactions.add(compactionTaskID);
             }
             else
             {
                 completedAncestors.addAll(ancestors);
             }
         }
+        cleanedUnfinishedCompactions.forEach(SystemKeyspace::finishCompaction);
 
         // remove old sstables from compactions that did complete
         for (Map.Entry<Descriptor, Set<Component>> sstableFiles : directories.sstableLister().list().entrySet())
         {
             Descriptor desc = sstableFiles.getKey();
             if (completedAncestors.contains(desc.generation))
             {
-                // if any of the ancestors were participating in a compaction, finish that compaction
-                logger.info("Going to delete leftover compaction ancestor {}", desc);
-                SSTable.delete(desc, sstableFiles.getValue());
-                UUID compactionTaskID = unfinishedCompactions.get(desc.generation);
-                if (compactionTaskID != null)
-                    SystemKeyspace.finishCompaction(unfinishedCompactions.get(desc.generation));
+                if (DRY_RUN_NON_COMPACTING_UNUSED_SSTABLE_CLEANUP && unfinishedCompactions.isEmpty())
+                {
+                    logger.warn("Would have deleted leftover compaction ancestor", UnsafeArg.of("desc", desc),
+                                SafeArg.of("keyspace", desc.ksname), SafeArg.of("cf", desc.cfname),
+                                SafeArg.of("generation", desc.generation), ancestorsArg);
+                } else
+                {
+                    // if any of the ancestors were participating in a compaction, finish that compaction
+                    logger.warn("Going to delete leftover compaction ancestor", UnsafeArg.of("desc", desc),
+                                SafeArg.of("keyspace", desc.ksname), SafeArg.of("cf", desc.cfname),
+                                SafeArg.of("generation", desc.generation), ancestorsArg);
+                    SSTable.delete(desc, sstableFiles.getValue());
+                    UUID compactionTaskID = unfinishedCompactions.get(desc.generation);
+                    if (compactionTaskID != null)
+                        SystemKeyspace.finishCompaction(unfinishedCompactions.get(desc.generation));
```

Comment:
nit
```suggestion
                    UUID compactionTaskID = unfinishedCompactions.get(desc.generation);
                    if (compactionTaskID != null)
                        SystemKeyspace.finishCompaction(compactionTaskID);
```
i don't know why the original code didn't already do this

or optionally
```suggestion
                    Optional.ofNullable(unfinishedCompactions.get(desc.generation)).ifPresent(SystemKeyspace::finishCompaction);
```